### PR TITLE
Assign functionality to start & stop buttons

### LINKED ISSUE #41: Create start model API call
**Problem:**
The models need to be initialised and run at the start of the simulation

**Proposed Solution:**
Determine the best way for this. Requires:
- [x] Check with Yue how the Opal model is started
- [x] There exists a start script - which requires the opal computer to be running, then [this script](https://github.com/yuezhu71/IDLES_python/blob/main/BalanceMechanism/BMStart.py) is run  (all the `RtlabApi` calls are what run the script)

Yue's code will check the datahub for a certain value to determine when to start the model. So this requires adding that value and endpoint to the datahub and changing that value from the vis system.

### PR DESCRIPTION
**Pull request by tsmbland**
# Description

This PR adds functionality to the start, stop and restart buttons in the control app.

Most of the changes can be found in `control.py`. This used to contain one large callback function (`update_button_click`) responsible for all the button callbacks (which were mostly empty), but since #92 Adrian has started moving button callbacks into their own functions which I think is a good move so I've followed suit here. I have written three new callbacks for the start, stop and restart buttons. The start and stop buttons are responsible for starting/stopping the live model, which it does via the new `start_model` and `stop_model` functions in `datahub_api.py`. These interact with the DataHub API introduced in [this PR](https://github.com/ImperialCollegeLondon/gridlington-datahub/pull/192/files), which modifies the DataHub variable `model_running` which is tracked by the model.  I hope I've done this correctly, however it's difficult for me to test without access to the live model. The start/stop buttons also interact with `data_interval` to start/stop data and figure updates in the vis system. The restart button sends `data_interval` back to the beginning, which resets the pre-set data. It already effectively did this when connected to the OVE (via `core.refresh_sections()`), but now it also works outside the OVE. In theory I think this button should also send a restart signal to the model via the DataHub, but it doesn't seem to me like this is possible yet, so I've left this as a TODO. (EDIT: Actually I _think_ stopping the model also restarts it, so this would just be a stop call followed by a start call - I will update)

I also had to rewrite some of the tests. The diffs look messy, but really I'm just moving the tests for each button into their own functions, and modifying tests for the start, stop and restart buttons in accordance with my changes.

Close #41 #61 #48

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (`python -m pytest`)
- [x] Pre-commit hooks run successfully (`pre-commit run --all-files`)

### CODE REVIEW

File: app/datahub_api.py [hunk 1]
```diff
@@ -94,3 +94,33 @@ def get_wesim_data() -> dict[str, dict]:  # type: ignore[type-arg]
...
+
+
+def start_model() -> str:
+    """Function for starting the model.
+
+    Returns:
+        str: Message to display on the control app
+    """
+    try:
+        if request_datahub("start").json():
```
Comment (dalonsoa) [inline app/datahub_api.py:106]: I'm not entirely sure, but I think this will return a dictionary, as JSON top object is always a dict, so shouldn't you be picking the key of interest out of that dict? Having said that, the [relevant endpoint in the datahub](https://github.com/ImperialCollegeLondon/gridlington-datahub/blob/4bae36ec857301ce07b4be62e0f5ce7eb2f86782/datahub/main.py#L315) just return a boolean flag, so I'm not entirely sure what you will have at this end...

Having said that, the name of this endpoint is misleading (a datahub issue) as it is not starting anything, just getting a flag about the status of the model. The one that actually starts the model - kind of, that one sets a flag - is `set_model_signals` which you use below.

File: app/datahub_api.py [hunk 2]
```diff
@@ -94,3 +94,33 @@ def get_wesim_data() -> dict[str, dict]:  # type: ignore[type-arg]
...
+
+
+def stop_model() -> str:
+    """Function for stopping the model.
+
+    Returns:
+        str: Message to display on the control app
+    """
+    try:
+        if request_datahub("stop").json():
```
Comment (dalonsoa) [inline app/datahub_api.py:121]: As above for `start`.

File: app/pages/control.py [hunk 3]
```diff
@@ -328,6 +301,82 @@ def default_button_click(n_clicks: int | None) -> list[str]:
...
+@callback(
+    [
+        Output("message", "children", allow_duplicate=True),
+        Output("data_interval", "n_intervals"),
+        Output("data_interval", "disabled", allow_duplicate=True),
+    ],
+    [Input("button_restart", "n_clicks")],
+    prevent_initial_call=True,
+)
+def restart_button_click(n_clicks: int | None) -> tuple[str, int, bool]:
```
Comment (dalonsoa) [inline app/pages/control.py:359]: Why is the `n_clicks` argument required?

File: app/pages/control.py [hunk 4]
```diff
@@ -241,10 +240,7 @@ def get_button(func: str, icon: str) -> html.Button:
     prevent_initial_call=True,
 )
 def update_button_click(
-    button_update: int | None,
-    button_start: int | None,
-    button_stop: int | None,
-    button_restart: int | None,
+    n_clicks: int | None,
```
Comment (dalonsoa) [inline app/pages/control.py:243]: What is the `n_clicks` argument used in the function for?

**Review by dalonsoa — APPROVED** [review]
I only have comments, not anything to change. 

Having said that, I am not entirely sure this will have any effect in the backend. The functions you have added connect to endpoints that, essentially, just set flags. But those flags, as far as I can tell, do nothing. They are not used anywhere else to define the behavior of the model. For example, the `reset_data` function in [here](https://github.com/ImperialCollegeLondon/gridlington-datahub/blob/4bae36ec857301ce07b4be62e0f5ce7eb2f86782/datahub/data.py#L15C11-L15C11), which presumably brings the data model back to its initial state, is not used. 

In summary, this PR looks good and is needed, but I'm not sure if it will work as intended without appropriate changes in the datahub.

**Comment by tsmbland** [comment]
> Having said that, I am not entirely sure this will have any effect in the backend. The functions you have added connect to endpoints that, essentially, just set flags. But those flags, as far as I can tell, do nothing. They are not used anywhere else to define the behavior of the model. For example, the `reset_data` function in [here](https://github.com/ImperialCollegeLondon/gridlington-datahub/blob/4bae36ec857301ce07b4be62e0f5ce7eb2f86782/datahub/data.py#L15C11-L15C11), which presumably brings the data model back to its initial state, is not used.

As discussed, I believe Yue's model will check the DataHub for the `model_running` flag to decide when to start/stop the model (see [here](https://github.com/ImperialCollegeLondon/gridlington-datahub/issues/183) and [here](https://github.com/ImperialCollegeLondon/gridlington-datahub/issues/184)). When stopping the model, I belive the model will then use the `signal_model_ready` API call to indicate to the DataHub when the model is ready to start again by setting the `model_resetting` flag. We probably need to use this flag in either the vis system or the DataHub to prevent the model from starting until it's ready, but this functionality doesn't appear to be in place without changes to the DataHub (as far as I can tell). Also, you're right that the `reset_data` function isn't used at all, and I imagine this should be included in one of the API calls.

I'm going to leave everything as it is for now and merge it, and will suggest that Adrian takes a look at this PR when he's back. I think it won't need any drastic changes, but Adrian knows how all the parts interact better than I do.

File: app/pages/control.py (hunk 3 above)
Comment (tsmbland) [reply app/pages/control.py:359]: This is used in the callback so that the function is run when n_clicks changes (i.e. when the button is pressed). It isn't used in the function, but needs to be there otherwise dash complains.

File: app/datahub_api.py (hunk 1 above)
Comment (tsmbland) [reply app/datahub_api.py:106]: > I'm not entirely sure, but I think this will return a dictionary, as JSON top object is always a dict, so shouldn't you be picking the key of interest out of that dict?

I've tested this with the DataHub running and it all seems to work as expected

**Comment by tsmbland** [comment]
@AdrianDAlessandro Can you check the start and stop buttons, and add the desired behaviour for the restart button? It all works when using the pre-set data, but not really sure what's required for the live model

**Comment by AdrianDAlessandro** [comment]
> @AdrianDAlessandro Can you check the start and stop buttons, and add the desired behaviour for the restart button? It all works when using the pre-set data, but not really sure what's required for the live model

Yes I checked them yesterday, they make the expected changes in the datahub, but the model needs to be developed to hook into the datahub, which is not up to us.

**Comment by tsmbland** [comment]
Ok. The restart button doesn't send any signals yet, just refreshes the page and resets the counter. Is this what we want?

**Comment by AdrianDAlessandro** [comment]
Yea, should be fine. I can't see a simple way to send a stop and then start signal that doesn't cause a race condition with the Opal model
